File: solutions/0731. My Calendar II/0731.cpp
// Time complexity: O(n)
// Space complexity: O(n)
class MyCalendarTwo {
 public:
  MyCalendarTwo() {}

  bool book(int start, int end) {
    for (const auto& time : overlapTimeline)
      if (doesOverlap(time.first, time.second, start, end))
        return false;

    for (const auto& time : timeline)
      if (doesOverlap(time.first, time.second, start, end))
        overlapTimeline.push_back(
          getOverlapped(time.first, time.second, start, end));

    timeline.push_back({start, end});
    return true;
  }

 private:
  vector<pair<int, int>> timeline;
  vector<pair<int, int>> overlapTimeline;

  bool doesOverlap(int start1, int end1, int start2, int end2) {
    return max(start1, start2) < min(end1, end2);
  }

  pair<int, int> getOverlapped(int start1, int end1, int start2, int end2) {
    return {max(start1, start2), min(end1, end2)};
  }
};

File: solutions/0731. My Calendar II/0731.py
# Time complexity: O(n)
# Space complexity: O(n)
class MyCalendarTwo:
  def __init__(self) -> None:
    self.timeline = []
    self.overlapTimeline = []

  def book(self, start: int, end: int) -> bool:
    for start1, end1 in self.overlapTimeline:
      if self._doesOverlap(start1, end1, start, end):
        return False

    for start1, end1 in self.timeline:
      if self._doesOverlap(start1, end1, start, end):
        self.overlapTimeline.append(
          self._getOverlapped(start1, end1, start, end)
        )

    self.timeline.append((start, end))
    return True

  def _doesOverlap(
    self, start1: int, end1: int, start2: int, end2: int
  ) -> bool:
    return max(start1, start2) < min(end1, end2)

  def _getOverlapped(
    self, start1: int, end1: int, start2: int, end2: int
  ) -> tuple:
    return max(start1, start2), min(end1, end2)